codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -239,22 +240,22 @@ private void writeGoAwayFrame(Http2GoAwayFrame frame, ChannelPromise promise) {
     private void writeStreamFrame(Http2StreamFrame frame, ChannelPromise promise) {
         if (frame instanceof Http2DataFrame) {
             Http2DataFrame dataFrame = (Http2DataFrame) frame;
-            http2Handler.encoder().writeData(http2HandlerCtx, frame.getStreamId(), dataFrame.content().retain(),
+            http2Handler.encoder().writeData(http2HandlerCtx, frame.streamId(), dataFrame.content().retain(),
                                              dataFrame.padding(), dataFrame.isEndStream(), promise);
         } else if (frame instanceof Http2HeadersFrame) {
             writeHeadersFrame((Http2HeadersFrame) frame, promise);
         } else if (frame instanceof Http2ResetFrame) {
             Http2ResetFrame rstFrame = (Http2ResetFrame) frame;
-            http2Handler.resetStream(http2HandlerCtx, frame.getStreamId(), rstFrame.errorCode(), promise);
+            http2Handler.resetStream(http2HandlerCtx, frame.streamId(), rstFrame.errorCode(), promise);
         } else {
             throw new UnsupportedMessageTypeException(frame);
         }
     }
 
     private void writeHeadersFrame(Http2HeadersFrame headersFrame, ChannelPromise promise) {
         final int streamId;
-        if (headersFrame.hasStreamId()) {
-            streamId = headersFrame.getStreamId();
+        if (isStreamIdValid(headersFrame.streamId())) {
```

Comment:
consider making `streamId` not final and assign it from the headers frame initially to avoid additional volatile reads